lib/string_decoder.js

```diff
@@ -282,3 +289,20 @@ function simpleWrite(buf) {
 function simpleEnd(buf) {
   return (buf && buf.length ? this.write(buf) : '');
 }
+
+function end(buf) {
+  let result = '';
+
+  if (this.encoding === 'utf16le')
+    result = utf16End.call(this, buf);
+  else if (this.encoding === 'utf8')
+    result = utf8End.call(this, buf);
+  else if (this.encoding === 'base64')
+    result = base64End.call(this, buf);
+  else
+    result = simpleEnd.call(this, buf);
+
+  this.reset();
```

Comment:
Would it be better to use `this.reset(this.encoding)` to avoid the `switch` in the constructor?

Reply:
We still cannot get rid of the `switch` in the constructor, right?

Reply:
Correct